### PR TITLE
feat: add missing ptype parameter for WatcherUpdatable methods

### PR DESCRIPTION
**Pull request by h7kanna**
Signed-off-by: Harsha Teja Kanna <h7kanna@gmail.com>

### CODE REVIEW

**Comment by h7kanna** [comment]
Actually  I was thinking why there is a separation of the watchers. Without the watcher update methods implemented by any watcher the implementation will be incomplete and difficult to maintain consistency.

**Comment by nodece** [comment]
LGTM, but it has breaking changes.

**Review by nodece — APPROVED** [review]

**Review by tangyang9464 — APPROVED** [review]
LGTM

**Comment by h7kanna** [comment]
I am sorry, just pushed some changes.
Please review again.

**Comment by h7kanna** [comment]
Hi, I have been working on Redis watcher, example of how this logic is used is here
https://github.com/h7kanna/redis-watcher/blob/watcher-batch-updates/callback.go#L95

File: internal_api.go [hunk 1]
```diff
@@ -336,28 +294,15 @@ func (e *Enforcer) updateFilteredPolicies(sec string, ptype string, newRules [][
...
+	ruleChanged, err := e.operator.UpdateFilteredPolicies(sec, ptype, oldRules, newRules)
+	if err != nil {
+		return ruleChanged, err
 	}
 
 	if e.watcher != nil && e.autoNotifyWatcher {
 		var err error
 		if watcher, ok := e.watcher.(persist.WatcherUpdatable); ok {
-			err = watcher.UpdateForUpdatePolicies(oldRules, newRules)
+			err = watcher.UpdateForUpdateFilteredPolicies(sec, ptype, oldRules, newRules)
```
Comment (closetool) [inline internal_api.go:305]: Why use `UpdateForUpdateFilteredPolicies`

File: persist/watcher_update.go [hunk 2]
```diff
@@ -19,8 +19,11 @@ type WatcherUpdatable interface {
...
 	// It is called after Enforcer.UpdatePolicy()
-	UpdateForUpdatePolicy(oldRule, newRule []string) error
+	UpdateForUpdatePolicy(sec, ptype string, oldRule, newRule []string) error
 	// UpdateForUpdatePolicies calls the update callback of other instances to synchronize their policy.
 	// It is called after Enforcer.UpdatePolicies()
-	UpdateForUpdatePolicies(oldRules, newRules [][]string) error
+	UpdateForUpdatePolicies(sec, ptype string, oldRules, newRules [][]string) error
+	// UpdateForUpdateFilteredPolicies calls the update callback of other instances to synchronize their policy.
+	// It is called after Enforcer.UpdateNamedFilteredPolicies()
+	UpdateForUpdateFilteredPolicies(sec, ptype string, oldRules, newRules [][]string) error
```
Comment (closetool) [inline persist/watcher_update.go:28]: This method is not needed. I suggest remove it.

**Review by closetool — CHANGES_REQUESTED** [review]
Plz fix it

File: persist/watcher_update.go (hunk 2 above)
Comment (h7kanna) [reply persist/watcher_update.go:28]: ok

Comment (h7kanna) [reply persist/watcher_update.go]: done

File: internal_api.go (hunk 1 above)
Comment (h7kanna) [reply internal_api.go]: Changed

Comment (h7kanna) [reply internal_api.go]: updated

**Comment by h7kanna** [comment]
Hi @closetool , Please let me know if I have to make any further changes or I have to remove anything for this PR.
I am looking to use this in the redis watcher enhancement I am working on.

**Comment by hsluoyz** [comment]
This change is too invasive. I don't think we need a PolicyOperator layer. There are already too many layers.

**Comment by h7kanna** [comment]
Hi @hsluoyz, Thanks for the review. Sorry for replying on the closed issue.
But do you think the methods in the internal_api can be made public to be reusable from watcher implementations.
Otherwise it is going to be code copy of the same functions. and they have to be in sync if casbin internal_api methods changes. I am reusing the internal_implementation here.
https://github.com/h7kanna/redis-watcher/blob/watcher-batch-updates/callback.go#L95

**Pull request closed without merging**